client/src/main/java/io/pravega/client/stream/impl/SegmentTransactionImpl.java

```diff
@@ -42,15 +83,22 @@ public UUID getId() {
 
     @Override
     public void flush() throws TxnFailedException {
+        checkFailed();
         try {
             out.flush();
+            synchronized (lock) {
+                removeCompleted();
+                checkFailed();
```

Comment:
`checkFailed` is inconsistently being invoked under a lock, see line 53.

Reply:
The lock is not needed. checkFailed uses txnFailedCause but it is an AtomicReference.

I cannot move this particular call outside the lock here because the precondition on the line below must be within the scope of the lock that performed removeCompleted() above and the checkfailed must occur between the two.